controllers/nifinodegroupautoscaler_controller.go

```diff
@@ -0,0 +1,428 @@
+	}
+
+	// Get the last configuration viewed by the operator.
+	o, err := patch.DefaultAnnotator.GetOriginalConfiguration(nodeGroupAutoscaler)
+	// Create it if not exist.
+	if o == nil {
+		if err := patch.DefaultAnnotator.SetLastAppliedAnnotation(nodeGroupAutoscaler); err != nil {
+			return RequeueWithError(r.Log, "could not apply last state to annotation", err)
+		}
+		if err := r.Client.Update(ctx, nodeGroupAutoscaler); err != nil {
+			return RequeueWithError(r.Log, "failed to update NifiNodeGroupAutoscaler", err)
+		}
+		o, err = patch.DefaultAnnotator.GetOriginalConfiguration(nodeGroupAutoscaler)
+	}
+
+	// Check if the cluster reference changed.
+	original := &v1alpha1.NifiNodeGroupAutoscaler{}
+	json.Unmarshal(o, original)
+	if !v1alpha1.ClusterRefsEquals([]v1alpha1.ClusterReference{original.Spec.ClusterRef, nodeGroupAutoscaler.Spec.ClusterRef}) {
+		nodeGroupAutoscaler.Spec.ClusterRef = original.Spec.ClusterRef
+	}
+
+	// lookup NifiCluster reference
+	// we do not want cached objects here. We want an accurate state of what the cluster is right now, so bypass the client cache by using the APIReader directly.
+	cluster := &v1alpha1.NifiCluster{}
+	err = r.APIReader.Get(ctx,
+		types.NamespacedName{
+			Name:      nodeGroupAutoscaler.Spec.ClusterRef.Name,
+			Namespace: nodeGroupAutoscaler.Spec.ClusterRef.Namespace,
+		},
+		cluster)
+	if err != nil {
+		return RequeueWithError(r.Log, fmt.Sprintf("failed to look up cluster reference %v+", nodeGroupAutoscaler.Spec.ClusterRef), err)
+	}
+
+	// Handle HorizontalAutoScaler
+	hpa, err := r.horizontalPodAutoscaler(r.Log, cluster, nodeGroupAutoscaler)
+	if err != nil {
+		return RequeueWithError(r.Log, "failed to generate horizontal pod autoscaler spec", err)
+	}
+	err = k8sutil.Reconcile(r.Log, r.Client, hpa, nil)
+	if err != nil {
+		return RequeueWithError(r.Log, "failed to reconcile horizontal pod autoscaler", err)
+	}
+
+	// Determine how many replicas there currently are and how many are desired for the appropriate node group
+	numDesiredReplicas := nodeGroupAutoscaler.Spec.Replicas
+	currentReplicas, maxId, err := r.getManagedNodes(nodeGroupAutoscaler, cluster.Spec.Nodes)
+	if err != nil {
+		return RequeueWithError(r.Log, "Failed to apply autoscaler node selector to cluster nodes", err)
+	}
+	numCurrentReplicas := int32(len(currentReplicas))
+
+	// if the current number of nodes being managed by this autoscaler is different than the replica setting,
+	// then set the autoscaler status to out of sync to indicate we're changing the NifiCluster node config
+	if numDesiredReplicas != numCurrentReplicas {
+		r.Log.Info(fmt.Sprintf("Replicas changed from %d to %d", numCurrentReplicas, numDesiredReplicas))
+		if err = r.updateAutoscalerReplicaState(ctx, nodeGroupAutoscaler, v1alpha1.AutoscalerStateOutOfSync); err != nil {
+			return RequeueWithError(r.Log, fmt.Sprintf("Failed to udpate node group autoscaler state for node group %s", nodeGroupAutoscaler.Spec.NodeConfigGroupId), err)
+		}
```

Comment:
Good catch! I think it would be sufficient to broaden the surrounding if-block to say:

```go
// If the replica setting doesn't match the node group size or if we're currently out of sync, then scale up/down
if numDesiredReplicas != numCurrentReplicas || autoscaler.Status.State == v1alpha1.AutoscalerStateOutOfSync {
  // ..... scale up/down .....
}
```

Do you agree?

Reply:
Yup :)